[BUG] AnthropicStreamingChatModel.builder().cacheTools creates cache block for each tool and easily exceeds Anthropic's limit of max 4 blocks with cache_control

<!-- Please provide as many details as possible, this will help us to deliver a fix as soon as possible. Thank you! -->

**Describe the bug**
<!-- A clear and concise description of what the bug is. -->
According to https://docs.anthropic.com/en/docs/build-with-claude/prompt-caching#best-practices-for-effective-caching, max 4 cache blocks can be created within messages send to Claude Sonnet (I am using 3.7), but cacheTools(true) configuration described at https://docs.langchain4j.dev/integrations/language-models/anthropic#caching leads to creating separate cache block for each tool specification in dev.langchain4j.model.anthropic.AnthropicStreamingChatModel#generate. This generate method calls AnthropicMapper.toAnthropicTools:

```
public static List<AnthropicTool> toAnthropicTools(List<ToolSpecification> toolSpecifications, AnthropicCacheType cacheToolsPrompt) {
        return toolSpecifications == null ? null : (List)toolSpecifications.stream().map((toolSpecification) -> toAnthropicTool(toolSpecification, cacheToolsPrompt)).collect(Collectors.toList());
}
```

And with more tools (blocks) than 4 (including also possible cached system message), this easily leads to an exception:

```
2025-05-17 13:08:19 117894 [OkHttp https://api.anthropic.com/...] ERROR c.e.a.t.c.r.AiChatStreamingResponseHandler - Error while processing AI response: {"type":"error","error":{"type":"invalid_request_error","message":"A maximum of 4 blocks with cache_control may be provided. Found 23."}}
dev.langchain4j.model.anthropic.internal.client.AnthropicHttpException: {"type":"error","error":{"type":"invalid_request_error","message":"A maximum of 4 blocks with cache_control may be provided. Found 23."}}
	at dev.langchain4j.model.anthropic.internal.client.DefaultAnthropicClient$1.onFailure(DefaultAnthropicClient.java:370)
	at okhttp3.internal.sse.RealEventSource.processResponse(RealEventSource.kt:52)
	at okhttp3.internal.sse.RealEventSource.onResponse(RealEventSource.kt:46)
	at okhttp3.internal.connection.RealCall$AsyncCall.run(RealCall.kt:519)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
```

**To Reproduce**
I am using the following model setup:

```
val chatModel = AnthropicStreamingChatModel.builder()
            .apiKey(chatModelConfig.getApiKey())
            .modelName("claude-3-7-sonnet-20250219")
            .temperature(0.3)
            .logRequests(true)
            .logResponses(true)
            .beta("prompt-caching-2024-07-31")
            .cacheSystemMessages(true)
            .cacheTools(true) // frequent tool definitions as "static" content
            .build()
```

**Expected behavior**
<!-- A clear and concise description of what you expected to happen. -->
I would assume that the tools will leverage the same caching block (or at least system message + max 3 tools will be cached, but that would be pitty not to cache all the available tools). I am interested to leverage my 22+ tools (from several MCP servers) and my system message cached without errors and use the possibilities of Anthropic API cost-effectively since it provides really interesting caching capabilities.

**Please complete the following information:**
- LangChain4j version: 1.0.0-beta3
- LLM(s) used: claude-3-7-sonnet-20250219
- Java version: openjdk 21.0.2 2024-01-16 LTS
OpenJDK Runtime Environment Temurin-21.0.2+13 (build 21.0.2+13-LTS)
- Spring Boot version (if applicable): N/A

**Additional context**
<!-- Add any other context about the problem here. -->

@Claudio-code First, I upgraded to LangChain4J 1.1.0, simulated the error again:
```
2025-07-11 10:57:17 46999 [ForkJoinPool.commonPool-worker-1] ERROR c.e.a.t.c.r.AiChatStreamingResponseHandler - Error while processing AI response: {"type":"error","error":{"type":"invalid_request_error","message":"A maximum of 4 blocks with cache_control may be provided. Found 15."}}
dev.langchain4j.exception.InvalidRequestException: {"type":"error","error":{"type":"invalid_request_error","message":"A maximum of 4 blocks with cache_control may be provided. Found 15."}}
        at dev.langchain4j.internal.ExceptionMapper$DefaultExceptionMapper.mapHttpStatusCode(ExceptionMapper.java:71)
        at dev.langchain4j.internal.ExceptionMapper$DefaultExceptionMapper.mapException(ExceptionMapper.java:44)
        at dev.langchain4j.model.anthropic.internal.client.DefaultAnthropicClient$1.onError(DefaultAnthropicClient.java:336)
        at dev.langchain4j.http.client.log.LoggingHttpClient$1.onError(LoggingHttpClient.java:75)
        at dev.langchain4j.http.client.jdk.JdkHttpClient.lambda$execute$0(JdkHttpClient.java:73)
        at dev.langchain4j.http.client.sse.ServerSentEventListenerUtils.ignoringExceptions(ServerSentEventListenerUtils.java:14)
        at dev.langchain4j.http.client.jdk.JdkHttpClient.lambda$execute$2(JdkHttpClient.java:73)
        at java.base/java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:718)
        at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
        at java.base/java.util.concurrent.CompletableFuture.postFire(CompletableFuture.java:614)
        at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:844)
        at java.base/java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:483)
        at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
        at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312)
        at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843)
        at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808)
        at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188)
Caused by: dev.langchain4j.exception.HttpException: {"type":"error","error":{"type":"invalid_request_error","message":"A maximum of 4 blocks with cache_control may be provided. Found 15."}}
        at dev.langchain4j.http.client.jdk.JdkHttpClient.lambda$execute$2(JdkHttpClient.java:72)
        ... 10 common frames omitted
```

then used your langchain4j-anthropic-1.2.0-SNAPSHOT.jar with fix and the results are fine, with expected cache_control blocks at the end of system messages and tool definition messages. Excerpt of code:

```
...
"system" : [ {
    "type" : "text",
    "cache_control" : {
      "type" : "ephemeral"
    },
    "text" : "..."
  } ]
...
...
{
    "name" : "fetch",
    "description" : "...",
    "input_schema" : {
      "type" : "object",
      "properties" : {
        "url" : {
          "type" : "string",
          "description" : "URL to fetch"
        },
        "max_length" : {
          "type" : "integer",
          "description" : "Maximum number of characters to return."
        },
        "start_index" : {
          "type" : "integer",
          "description" : "On return output starting at this character index, useful if a previous fetch was truncated and more context is required." 
        },
        "raw" : {
          "type" : "boolean",
          "description" : "Get the actual HTML content of the requested page, without simplification."
        }
      },
      "required" : [ "url" ]
    }
  }, {
    "name" : "brave_local_search",
    "description" : "...",
    "input_schema" : {
      "type" : "object",
      "properties" : {
        "query" : {
          "type" : "string",
          "description" : "Local search query (e.g. 'pizza near Central Park')"
        },
        "count" : {
          "type" : "number",
          "description" : "Number of results (1-20, default 5)"
        }
      },
      "required" : [ "query" ]
    }
  }, {
    "name" : "resolve-library-id",
    "description" : "...",
    "input_schema" : {
      "type" : "object",
      "properties" : {
        "libraryName" : {
          "type" : "string",
          "description" : "Library name to search for and retrieve a Context7-compatible library ID."
        }
      },
      "required" : [ "libraryName" ]
    }
  }, {
    "name" : "get-library-docs",
    "description" : "...",
    "input_schema" : {
      "type" : "object",
      "properties" : {
        "context7CompatibleLibraryID" : {
          "type" : "string",
          "description" : "Exact Context7-compatible library ID (e.g., '/mongodb/docs', '/vercel/next.js', '/supabase/supabase', '/vercel/next.js/v14.3.0-canary.87') retrieved from 'resolve-library-id' or directly from user query in the format '/org/project' or '/org/project/version'."
        },
        "topic" : {
          "type" : "string",
          "description" : "Topic to focus documentation on (e.g., 'hooks', 'routing')."
        },
        "tokens" : {
          "type" : "number",
          "description" : "Maximum number of tokens of documentation to retrieve (default: 10000). Higher values provide more context but consume more tokens."
        }
      },
      "required" : [ "context7CompatibleLibraryID" ]
    }
  }, {
    "name" : "brave_web_search",
    "description" : "...",
    "input_schema" : {
      "type" : "object",
      "properties" : {
        "query" : {
          "type" : "string",
          "description" : "Search query (max 400 chars, 50 words)"
        },
        "count" : {
          "type" : "number",
          "description" : "Number of results (1-20, default 10)"
        },
        "offset" : {
          "type" : "number",
          "description" : "Pagination offset (max 9, default 0)"
        }
      },
      "required" : [ "query" ]
    },
    "cache_control" : {
      "type" : "ephemeral"
    }
  } ]
}
```

And I also see some tokens cached and spared from AnthropicTokenUsage:

`2025-07-11 11:00:34 78284 [ForkJoinPool.commonPool-worker-6] INFO  c.e.a.tools.chat.stats.ContextStats - Tokens: 8947 input, 550 output, 0 cache creation input, 5397 cache read input
`

So, good job! That means a lot! Your fix can be merged.

--- a/langchain4j-anthropic/src/main/java/dev/langchain4j/model/anthropic/internal/mapper/AnthropicMapper.java
+++ b/langchain4j-anthropic/src/main/java/dev/langchain4j/model/anthropic/internal/mapper/AnthropicMapper.java
@@ -55,6 +55,7 @@ import dev.langchain4j.model.chat.request.json.JsonObjectSchema;
 import dev.langchain4j.model.output.FinishReason;
 import dev.langchain4j.model.output.TokenUsage;
 import java.util.ArrayList;
+import java.util.Collections;
 import java.util.List;
 import java.util.Map;
 
@@ -160,7 +161,7 @@ public class AnthropicMapper {
                 .map(message -> (SystemMessage) message)
                 .collect(toList());
 
-        SystemMessage lastSystemMessage = systemMessages.get(systemMessages.size() - 1);
+        SystemMessage lastSystemMessage = systemMessages.isEmpty() ? null : systemMessages.get(systemMessages.size() - 1);
         return systemMessages.stream()
                 .map(message -> {
                     boolean isLastItem = message.equals(lastSystemMessage);
@@ -223,7 +224,7 @@ public class AnthropicMapper {
 
     public static List<AnthropicTool> toAnthropicTools(
             List<ToolSpecification> toolSpecifications, AnthropicCacheType cacheToolsPrompt) {
-        ToolSpecification lastToolSpecification = toolSpecifications.get(toolSpecifications.size() - 1);
+        ToolSpecification lastToolSpecification = toolSpecifications.isEmpty() ? null : toolSpecifications.get(toolSpecifications.size() - 1);
         return toolSpecifications.stream()
                 .map(toolSpecification -> {
                     boolean isLastItem = toolSpecification.equals(lastToolSpecification);

--- a/langchain4j-anthropic/src/main/java/dev/langchain4j/model/anthropic/internal/mapper/AnthropicMapper.java
+++ b/langchain4j-anthropic/src/main/java/dev/langchain4j/model/anthropic/internal/mapper/AnthropicMapper.java
@@ -1,6 +1,7 @@
 package dev.langchain4j.model.anthropic.internal.mapper;
 
 import static dev.langchain4j.internal.Exceptions.illegalArgument;
+import static dev.langchain4j.internal.JsonSchemaElementUtils.toMap;
 import static dev.langchain4j.internal.Utils.isNotNullOrBlank;
 import static dev.langchain4j.internal.Utils.isNullOrBlank;
 import static dev.langchain4j.internal.Utils.isNullOrEmpty;
@@ -9,7 +10,6 @@ import static dev.langchain4j.model.anthropic.internal.api.AnthropicContentBlock
 import static dev.langchain4j.model.anthropic.internal.api.AnthropicContentBlockType.TOOL_USE;
 import static dev.langchain4j.model.anthropic.internal.api.AnthropicRole.ASSISTANT;
 import static dev.langchain4j.model.anthropic.internal.api.AnthropicRole.USER;
-import static dev.langchain4j.internal.JsonSchemaElementUtils.toMap;
 import static dev.langchain4j.model.anthropic.internal.client.Json.fromJson;
 import static dev.langchain4j.model.anthropic.internal.client.Json.toJson;
 import static dev.langchain4j.model.output.FinishReason.LENGTH;
@@ -55,7 +55,6 @@ import dev.langchain4j.model.chat.request.json.JsonObjectSchema;
 import dev.langchain4j.model.output.FinishReason;
 import dev.langchain4j.model.output.TokenUsage;
 import java.util.ArrayList;
-import java.util.Collections;
 import java.util.List;
 import java.util.Map;
 
@@ -116,7 +115,8 @@ public class AnthropicMapper {
                                 ensureNotBlank(image.base64Data(), "base64Data"));
                     } else if (content instanceof PdfFileContent pdfFileContent) {
                         PdfFile pdfFile = pdfFileContent.pdfFile();
-                        return new AnthropicPdfContent(pdfFile.mimeType(), ensureNotBlank(pdfFile.base64Data(), "base64Data"));
+                        return new AnthropicPdfContent(
+                                pdfFile.mimeType(), ensureNotBlank(pdfFile.base64Data(), "base64Data"));
                     } else {
                         throw illegalArgument("Unknown content type: " + content);
                     }
@@ -161,7 +161,8 @@ public class AnthropicMapper {
                 .map(message -> (SystemMessage) message)
                 .collect(toList());
 
-        SystemMessage lastSystemMessage = systemMessages.isEmpty() ? null : systemMessages.get(systemMessages.size() - 1);
+        SystemMessage lastSystemMessage =
+                systemMessages.isEmpty() ? null : systemMessages.get(systemMessages.size() - 1);
         return systemMessages.stream()
                 .map(message -> {
                     boolean isLastItem = message.equals(lastSystemMessage);
@@ -224,7 +225,8 @@ public class AnthropicMapper {
 
     public static List<AnthropicTool> toAnthropicTools(
             List<ToolSpecification> toolSpecifications, AnthropicCacheType cacheToolsPrompt) {
-        ToolSpecification lastToolSpecification = toolSpecifications.isEmpty() ? null : toolSpecifications.get(toolSpecifications.size() - 1);
+        ToolSpecification lastToolSpecification =
+                toolSpecifications.isEmpty() ? null : toolSpecifications.get(toolSpecifications.size() - 1);
         return toolSpecifications.stream()
                 .map(toolSpecification -> {
                     boolean isLastItem = toolSpecification.equals(lastToolSpecification);
@@ -260,10 +262,11 @@ public class AnthropicMapper {
             return null;
         }
 
-        AnthropicToolChoiceType toolChoiceType = switch (toolChoice) {
-            case AUTO -> AnthropicToolChoiceType.AUTO;
-            case REQUIRED -> AnthropicToolChoiceType.ANY;
-        };
+        AnthropicToolChoiceType toolChoiceType =
+                switch (toolChoice) {
+                    case AUTO -> AnthropicToolChoiceType.AUTO;
+                    case REQUIRED -> AnthropicToolChoiceType.ANY;
+                };
 
         return AnthropicToolChoice.from(toolChoiceType);
     }

--- a/langchain4j-anthropic/src/main/java/dev/langchain4j/model/anthropic/internal/mapper/AnthropicMapper.java
+++ b/langchain4j-anthropic/src/main/java/dev/langchain4j/model/anthropic/internal/mapper/AnthropicMapper.java
@@ -160,9 +160,10 @@ public class AnthropicMapper {
                 .map(message -> (SystemMessage) message)
                 .collect(toList());
 
+        SystemMessage lastSystemMessage = systemMessages.get(systemMessages.size() - 1);
         return systemMessages.stream()
                 .map(message -> {
-                    boolean isLastItem = message.equals(systemMessages.get(systemMessages.size() - 1));
+                    boolean isLastItem = message.equals(lastSystemMessage);
                     if (isLastItem && cacheType != AnthropicCacheType.NO_CACHE) {
                         return new AnthropicTextContent(message.text(), cacheType.cacheControl());
                     }
@@ -222,9 +223,10 @@ public class AnthropicMapper {
 
     public static List<AnthropicTool> toAnthropicTools(
             List<ToolSpecification> toolSpecifications, AnthropicCacheType cacheToolsPrompt) {
+        ToolSpecification lastToolSpecification = toolSpecifications.get(toolSpecifications.size() - 1);
         return toolSpecifications.stream()
                 .map(toolSpecification -> {
-                    boolean isLastItem = toolSpecification.equals(toolSpecifications.get(toolSpecifications.size() - 1));
+                    boolean isLastItem = toolSpecification.equals(lastToolSpecification);
                     if (isLastItem && cacheToolsPrompt != AnthropicCacheType.NO_CACHE) {
                         return toAnthropicTool(toolSpecification, cacheToolsPrompt);
                     }

--- a/langchain4j-anthropic/src/main/java/dev/langchain4j/model/anthropic/internal/mapper/AnthropicMapper.java
+++ b/langchain4j-anthropic/src/main/java/dev/langchain4j/model/anthropic/internal/mapper/AnthropicMapper.java
@@ -155,14 +155,18 @@ public class AnthropicMapper {
 
     public static List<AnthropicTextContent> toAnthropicSystemPrompt(
             List<ChatMessage> messages, AnthropicCacheType cacheType) {
-        return messages.stream()
+        List<SystemMessage> systemMessages = messages.stream()
                 .filter(message -> message instanceof SystemMessage)
+                .map(message -> (SystemMessage) message)
+                .collect(toList());
+
+        return systemMessages.stream()
                 .map(message -> {
-                    SystemMessage systemMessage = (SystemMessage) message;
-                    if (cacheType != AnthropicCacheType.NO_CACHE) {
-                        return new AnthropicTextContent(systemMessage.text(), cacheType.cacheControl());
+                    boolean isLastItem = message.equals(systemMessages.get(systemMessages.size() - 1));
+                    if (isLastItem && cacheType != AnthropicCacheType.NO_CACHE) {
+                        return new AnthropicTextContent(message.text(), cacheType.cacheControl());
                     }
-                    return new AnthropicTextContent(systemMessage.text());
+                    return new AnthropicTextContent(message.text());
                 })
                 .collect(toList());
     }
@@ -219,7 +223,13 @@ public class AnthropicMapper {
     public static List<AnthropicTool> toAnthropicTools(
             List<ToolSpecification> toolSpecifications, AnthropicCacheType cacheToolsPrompt) {
         return toolSpecifications.stream()
-                .map(toolSpecification -> toAnthropicTool(toolSpecification, cacheToolsPrompt))
+                .map(toolSpecification -> {
+                    boolean isLastItem = toolSpecification.equals(toolSpecifications.get(toolSpecifications.size() - 1));
+                    if (isLastItem && cacheToolsPrompt != AnthropicCacheType.NO_CACHE) {
+                        return toAnthropicTool(toolSpecification, cacheToolsPrompt);
+                    }
+                    return toAnthropicTool(toolSpecification, AnthropicCacheType.NO_CACHE);
+                })
                 .collect(toList());
     }
 

